### ISSUE TITLE
hwmon-path will change after restart

### ISSUE THREAD

**Issue by ati46** [polybar/polybar#2078]
After restarting the system, the Temperature Module is no longer displayed, 
use
 for i in / sys / class / hwmon / hwmon * / temp * _input; do echo "$ (<$ (dirname $ i) / name): $ (cat $ {i% _ *} _ label 2> / dev / null || echo $ (basename $ {i% _ *})) $ (readlink -f $ i) "; done 
command found that the path is different from before the system was restarted.
Not every time you restart.

**Comment by patrick96** [comment]
Hmm, I'm not sure why the script on the wiki tells people to use the path without symlinks instead of the path to the file inside `/sys/class/hwmon/`.

Could you try running the following command and using the paths from there: 

```bash
for i in /sys/class/hwmon/hwmon*/temp*_input; do 
  echo "$(<$(dirname $i)/name): $(cat ${i%_*}_label 2>/dev/null || echo $(basename ${i%_*})) $i";
done
```

**Comment by ati46** [comment]
Thank you very much, this is currently working for me.
After I use it for a few days, I will again report whether the problem has changed after restarting.
@patrick96

**Comment by patrick96** [comment]
Alright, thanks for reporting.

I will change the wiki entry if this indeed works.

**Comment by ati46** [comment]
The output will still change after restart.
The data changes randomly between 0,1, 2, 3, 4

Last start result：
`
$ for i in /sys/class/hwmon/hwmon*/temp*_input; do 
for>   echo "$(<$(dirname $i)/name): $(cat ${i%_*}_label 2>/dev/null || echo $(basename ${i%_*})) $i";
for> done
nvme: Composite /sys/class/hwmon/hwmon0/temp1_input
nvme: Sensor 1 /sys/class/hwmon/hwmon0/temp2_input
nvme: Sensor 2 /sys/class/hwmon/hwmon0/temp3_input
coretemp: Package id 0 /sys/class/hwmon/hwmon3/temp1_input
coretemp: Core 0 /sys/class/hwmon/hwmon3/temp2_input
coretemp: Core 1 /sys/class/hwmon/hwmon3/temp3_input
coretemp: Core 2 /sys/class/hwmon/hwmon3/temp4_input
coretemp: Core 3 /sys/class/hwmon/hwmon3/temp5_input
iwlwifi_1: temp1 /sys/class/hwmon/hwmon4/temp1_input
`

The result of this launch：
$ for i in /sys/class/hwmon/hwmon*/temp*_input; do 
for>   echo "$(<$(dirname $i)/name): $(cat ${i%_*}_label 2>/dev/null || echo $(basename ${i%_*})) $i";
for> done
nvme: Composite /sys/class/hwmon/hwmon0/temp1_input
nvme: Sensor 1 /sys/class/hwmon/hwmon0/temp2_input
nvme: Sensor 2 /sys/class/hwmon/hwmon0/temp3_input
coretemp: Package id 0 /sys/class/hwmon/hwmon4/temp1_input
coretemp: Core 0 /sys/class/hwmon/hwmon4/temp2_input
coretemp: Core 1 /sys/class/hwmon/hwmon4/temp3_input
coretemp: Core 2 /sys/class/hwmon/hwmon4/temp4_input
coretemp: Core 3 /sys/class/hwmon/hwmon4/temp5_input
iwlwifi_1: temp1 /sys/class/hwmon/hwmon2/temp1_input
`

**Comment by patrick96** [comment]
That's too bad.

I guess the only way to solve this is to automatically set the right path on startup in an environment variable and then include that in polybar like so:

```dosini
hwmon-path = ${env:HWMON_PATH}
```

The script that launches polybar could then do something like this:

```bash
for i in /sys/class/hwmon/hwmon*/temp*_input; do 
    if [ "$(<$(dirname $i)/name): $(cat ${i%_*}_label 2>/dev/null || echo $(basename ${i%_*}))" = "coretemp: Core 0" ]; then
        export HWMON_PATH="$i"
    fi
done
```

**Comment by kevariable** [comment]
i had the same issue, can be done?

**Comment by kevariable** [comment]
for i in /sys/class/hwmon/hwmon*/temp*_input; do
  echo "$(<$(dirname $i)/name): $(cat ${i%_*}_label 2>/dev/null || echo $(basename ${i%_*})) $i";
done
k10temp: temp1 /sys/class/hwmon/hwmon3/temp1_input
amdgpu: edge /sys/class/hwmon/hwmon4/temp1_input

can you help me

**Comment by patrick96** [comment]
@kevariable I showd the workaround in my previous comment. You have to replace `coretemp: Core 0` with the name of the sensor, so in your case either `k10temp: temp1` or `amdgpu: edge`

**Comment by jwhendy** [comment]
Not sure this is worth opening a separate issue for, but I found what I believe to be this path changing upon resume from suspend (not hibernate).

At present, I run the above script and have `HWMON_PATH` set via `i3` when I `startx`. If this path isn't stable through a resume, though, would I have to tie this to each lid opening event? Alternatively, if I set this path setting script to a cron job... would polybar pick up the change on the fly, or only when it starts?

Basically, what event(s) trigger polybar to read `${env:HWMON_PATH}`?

**Comment by patrick96** [comment]
Polybar reads the environment variables when it starts. Also there isn't a general way to externally change the environment variables of a process (they are inherited from the parent process).

I would suggest creating a launch script for polybar and setting the path in there before you start polybar. This makes sure polybar has access to the variable, then you just have to run the script whenever you think the path has changed.

**Comment by jwhendy** [comment]
Thanks @patrick96. If I understand correctly, I create a new script that combines my currently separate actions: a) find and set `HWMON_PATH` and b) launch polybar. Then I run this script upon resume from suspend.

(Also, this script also needs to kill the current polybar instance, right?)

**Comment by patrick96** [comment]
> (Also, this script also needs to kill the current polybar instance, right?)

Yes. In general it's a good idea if your launch script terminates any running polybar instances.
